[Lens] Manual Annotations

mbondyra opened the pull request:
## Summary
Introduces new type of the layer for xy chart: annotations.
Only targets manual, 'point in time' annotations.

The icon set is covering most cases for tsvb (to be able to migrate in the future) and all we use for reference lines 
Ranges will be added soon (in separate pr). 
Drag and drop will be added soon (in separate pr).
Tooltip collect the info for all the annotations from the interval. 

Default look:
<img width="217" alt="Screenshot 2022-03-23 at 09 47 50" src="https://user-images.githubusercontent.com/4283304/159659443-2af03847-6d4d-44a0-98db-5b7c5babf477.png">
Grouping and custom annotations:
<img width="969" alt="Screenshot 2022-03-23 at 09 49 27" src="https://user-images.githubusercontent.com/4283304/159659744-c49bed46-4ba8-4048-8d61-95d3cadf4e97.png">
<img width="439" alt="Screenshot 2022-03-23 at 09 46 57" src="https://user-images.githubusercontent.com/4283304/159659363-52d673f5-6738-410a-b737-2bfba01f131d.png">
<img width="394" alt="Screenshot 2022-03-23 at 09 47 05" src="https://user-images.githubusercontent.com/4283304/159659369-cdf7174c-ddd1-4afd-9490-323dd6fc1e5e.png">


#### Event annotations service

At the moment the service only contains the expressions.

mbondyra commented:
@elasticmachine merge upstream

elasticmachine commented:
Pinging @elastic/kibana-vis-editors @elastic/kibana-vis-editors-external (Team:VisEditors)

jbudz reviewed (approved):
limits.yml LGTM

mbondyra commented on src/plugins/event_annotation/public/event_annotation_service/service.tsx:
```diff
@@ -0,0 +1,54 @@
+ */
+
+// @ts-ignore
+import chroma from 'chroma-js';
```
I have no idea how it got here 😅 thank you!

MichaelMarcialis reviewed (commented):
Excellent work here, @mbondyra! Leaving you a few comments/suggestions below for your review. Let me know if you have any questions.

- There's a stray `EuiSpacer` component that currently appears at the end of the annotation layer panel's header. Can we remove?

![image](https://user-images.githubusercontent.com/3884767/159337878-89e72f80-5d70-478f-ad8b-379fa2deddfb.png)

- Can we add a `responsive={false}` prop on the `EuiFlexGroup` used to contain the inputs for the `Line` form row? At small viewport sizes, having it be responsive causes the button group to break out onto its own line, which I'd like to avoid.

![image](https://user-images.githubusercontent.com/3884767/159338485-9392ee7c-a5ef-4f77-88b4-a7f1d638d17e.png) 

- The placement and appearance portions of form inputs appear to be grouped together in a single section within the annotation configuration flyout. Would it be possible to group these form elements into two separate sections, similar to what is shown in the wireframes (and other Lens configuration flyouts), with a border dividing the sections?

![image](https://user-images.githubusercontent.com/3884767/159339388-c7a36de5-b6cf-430b-a185-54f7a5bf0135.png)

![image](https://user-images.githubusercontent.com/3884767/159339070-8dc800f9-a02f-45d7-9e25-18ae6a807eee.png)

- Related to the previous comment, the `Placement` and `Appearance` headers suggested in the wireframes for the annotation configuration flyout form sections are missing. I know that we previously discussed that it wouldn't be simple to add such headings in other configuration flyouts across the Lens experience, but we might be able to add such headings in the annotation configuration flyout as a first step (and then add similar headers to the other flyouts iteratively). Was that explored for this PR, or would that be something for a separate PR?

- I noticed that annotation-specific tooltip content is only visible when hovering on that annotation's text or icon decoration. Hovering the annotation line does not include this annotation content in the tooltip. Additionally, if text or no icon decoration is being shown, there is no way to view the annotation content in a tooltip. Is this intentional? CCing @gvnmagni in case he has additional comments.

- Related to the previous comment, annotation-specific tooltips aren't show the correct/exact datetime of the annotation. Instead, the datetime they show is the beginning of the bar interval (when using bar vertical stacked). Is that intentional? I feel like we should be showing the exact datetime that the user has annotated. See the screenshot below with the inconsistency with the annotation tooltip datetime and configuration. CCing @gvnmagni in case he has additional comments.

![image](https://user-images.githubusercontent.com/3884767/159356917-5fdde677-12c2-4aae-8394-b37bde4c454d.png)

mbondyra commented on x-pack/plugins/lens/public/editor_frame_service/editor_frame/config_panel/layer_panel.tsx:
```diff
@@ -180,16 +182,18 @@ export function LayerPanel(
+            groupId,
+            dropType,
+          })
+        : true;
```
Yes - this code will never execute for now because nothing is returned for `getDropProps`, but you're right 👌🏼

mbondyra commented on x-pack/plugins/lens/public/xy_visualization/annotations/expression.tsx:
```diff
@@ -0,0 +1,231 @@
+  customTooltipDetails?: AnnotationTooltipFormatter | undefined;
+}
+
+const groupVisibleConfigsByInterval = (
```
To do so, we'll need to have data layer interval and one of the timestamp from data layer known... I am not sure if we can do it? Let's discuss :)

mbondyra commented on x-pack/plugins/lens/public/xy_visualization/annotations/config_panel/index.tsx:
```diff
@@ -0,0 +1,171 @@
+}) => {
+  return (
+    <EuiFormRow display="rowCompressed" fullWidth label={label}>
+      <EuiDatePicker
+        showTimeSelect
+        selected={value}
+        onChange={onChange}
+        dateFormat="MMM D, YYYY @ HH:mm:ss.SSS"
+        data-test-subj="lnsXY_annotation_date_picker"
+      />
```
Yeah, for now I'd ignore this issue. Hopefully it will be solved soon. 🤞🏼

mbondyra commented:
Thanks @flash1293 and @MichaelMarcialis for reviews. Regarding Michael's comments, I've corrected all the online resolved ones. About the grouped comments:

#### To be discussed:
> - I noticed that annotation-specific tooltip content is only visible when hovering on that annotation's text or icon decoration. Hovering the annotation line does not include this annotation content in the tooltip. Additionally, if text or no icon decoration is being shown, there is no way to view the annotation content in a tooltip. Is this intentional? CCing @gvnmagni in case he has additional comments.

I think elastic charts don't give us this possibility? CC @markov00 Same problem exists for annotations for TSVB - if we want to address it, probably we won't in this release.

#### Resolved

> - There's a stray `EuiSpacer` component that currently appears at the end of the annotation layer panel's header. Can we remove?

Done!

> - Can we add a `responsive={false}` prop on the `EuiFlexGroup` ... 

Done!

> - The placement and appearance portions of form inputs appear to be grouped together in a single section within the annotation configuration flyout. Would it be possible to group these form elements into two separate sections, similar to what is shown in the wireframes (and other Lens configuration flyouts), with a border dividing the sections?

Done for annotations: 
<img width="505" alt="Screenshot 2022-03-21 at 14 00 15" src="https://user-images.githubusercontent.com/4283304/159452204-e70f28cd-824e-4a34-9dd5-4d283c71b52b.png">




> - Related to the previous comment, annotation-specific tooltips aren't showing the correct/exact datetime of the annotation. 

That's a bug, fixed, thank you!

stratoula commented on src/plugins/event_annotation/common/event_annotation_group/index.ts:
```diff
@@ -0,0 +1,47 @@
+    args: {
+      annotations: {
+        types: ['manual_event_annotation'],
+        help: 'Annotation configs',
```
nit: Should this be localized?

stratoula commented on src/plugins/event_annotation/common/manual_event_annotation/index.ts:
```diff
@@ -0,0 +1,68 @@
+  args: {
+    time: {
+      types: ['string'],
+      help: 'Timestamp for annotation',
```
It will be good if we localized all these texts.

stratoula commented on src/plugins/event_annotation/kibana.json:
```diff
@@ -0,0 +1,16 @@
+  "extraPublicDirs": [
+    "common"
+  ],
+  "requiredPlugins": [
```
I think that it will be good to also add the description here. I think it is connected with our documentation.

stratoula commented on x-pack/plugins/lens/public/xy_visualization/xy_suggestions.ts:
```diff
@@ -521,7 +521,9 @@ function buildSuggestion({
         // Remove layers that aren't being suggested
-        .filter((layer) => keptLayerIds.includes(layer.layerId))
+        .filter(
+          (layer) => keptLayerIds.includes(layer.layerId) || layer.layerType === 'annotations'
```
nit: we use the constant here

stratoula reviewed (commented):
Amazing Job Marta 👏 💯 I am leaving some very small comments from the code review. I will proceed with testing it later today.

MichaelMarcialis commented on x-pack/plugins/lens/public/xy_visualization/xy_config_panel/shared/line_style_settings.tsx:
```diff
@@ -40,7 +40,7 @@ export const LineStyleSettings = ({
         })}
       >
-        <EuiFlexGroup gutterSize="s" justifyContent="spaceBetween">
+        <EuiFlexGroup gutterSize="s" justifyContent="spaceBetween" responsive={false}>
```
The line style `EuiButtonGroup` that is a descendant of this flex group should have its `isFullWidth` prop removed. Also, the `EuiButtonGroup`'s parent `EuiFlexItem` should have it's `grow` prop set to `{false}`. This will ensure the button group never grows bigger than the size of its contents (i.e. the icons).

MichaelMarcialis commented on x-pack/plugins/lens/public/xy_visualization/xy_config_panel/shared/line_style_settings.tsx:
```diff
@@ -40,7 +40,7 @@ export const LineStyleSettings = ({
       >
-        <EuiFlexGroup gutterSize="s" justifyContent="spaceBetween">
+        <EuiFlexGroup gutterSize="s" justifyContent="spaceBetween" responsive={false}>
           <EuiFlexItem grow={true}>
```
I believe the `grow={true}` prop can be removed here, as this is the default for the component.

```suggestion
          <EuiFlexItem>
```

MichaelMarcialis commented on x-pack/plugins/lens/public/xy_visualization/annotations/config_panel/index.tsx:
```diff
@@ -0,0 +1,192 @@
+        />
+      </DimensionEditorSection>
+      <DimensionEditorSection hasBorder>
+        <EuiTitle size="xs" className="lnsXyConfigHeading">
```
To avoid using the same font size as the flyout header and improve the visual hierarchy, can we take the font size down one step for this title?

```suggestion
        <EuiTitle size="xxs" className="lnsXyConfigHeading">
```

MichaelMarcialis commented on x-pack/plugins/lens/public/xy_visualization/annotations/config_panel/index.tsx:
```diff
@@ -0,0 +1,192 @@
+  return (
+    <>
+      <DimensionEditorSection>
+        <EuiTitle size="xs" className="lnsXyConfigHeading">
```
To avoid using the same font size as the flyout header and improve the visual hierarchy, can we take the font size down one step for this title?

```suggestion
        <EuiTitle size="xxs" className="lnsXyConfigHeading">
```

MichaelMarcialis reviewed (approved):
Thanks so much for making those changes, @mbondyra. You've done an amazing job, as always. Well done!

I've left you a few small comments below, but assuming these get addressed and pending any other details that come from our meeting today, this looks great to me. Approving.

alexwizp commented on src/plugins/event_annotation/public/event_annotation_service/index.tsx:
```diff
@@ -0,0 +1,20 @@
+import { EventAnnotationServiceType } from './types';
+
+export class EventAnnotationService {
+  private eventAnnotationService: EventAnnotationServiceType | undefined = undefined;
```
nit: 
`private eventAnnotationService?: EventAnnotationServiceType`;

mbondyra replied:
```suggestion
  private eventAnnotationService?: EventAnnotationServiceType;
```

alexwizp commented on src/plugins/event_annotation/public/event_annotation_service/service.tsx:
```diff
@@ -0,0 +1,49 @@
+ */
+
+import { EventAnnotationServiceType } from './types';
+import { defaultAnnotationColor } from '..';
```
It seems to me that it is not very correct to do such an import, there is clearly a cross reference here.  

With tree shaking, in such cases a duplicate is created and this leads in some cases to errors

alexwizp commented on src/plugins/event_annotation/public/event_annotation_service/helpers.ts:
```diff
@@ -0,0 +1,9 @@
+ * Side Public License, v 1.
+ */
+import { euiLightVars } from '@kbn/ui-theme';
+export const defaultAnnotationColor = euiLightVars.euiColorAccent;
```
nit: I probably recommended not to fix the value at the time of import. Maybe `geDefaultAnnotationColor = () => euiLightVars.euiColorAccent;`

mbondyra replied:
Why does it matter? 🤔

markov00 replied:
@mbondyra not related to alex suggestion, but to the color: the `euiColorAccent` differs if you are in dark mode or light mode. in the light theme the color is  #F04E98 but is  #F68FBE in dark mode.
@MichaelMarcialis ideas?

mbondyra replied:
@markov00 does it matter? Why is it concerning?

flash1293 replied:
On switching to dark mode the annotations will still have the light theme color because it's putting the hex value into the saved object. I think that is OK for this PR, it's the same for all of the place where we allow the users to pick a color (series, reference lines and so on). When we discussed this the last time we said we would tackle this problem separately in a general way and apply the solutions to all the places which allow this kind of thing right now.

MichaelMarcialis replied:
> @mbondyra not related to alex suggestion, but to the color: the `euiColorAccent` differs if you are in dark mode or light mode. in the light theme the color is `#F04E98` but is `#F68FBE` in dark mode.
> @MichaelMarcialis ideas?

@markov00: That's a good point; something we hadn't considered in our discussion yesterday. Ideally, we would have a single color value that works well for both light and dark modes. Otherwise, the default color is not a purely static value, but a partially dynamic one that changes depending on the theme mode selected when the visualization was created. This could be unexpected or undesirable to users (especially after changing an existing space's theme mode, when sharing/copying visualization across spaces, and in the future when users can set their own theme modes). This is part of the reason why we chose a gray color for the default reference line color, because it works in both theme modes.

In this case, the difference in the color value between light and dark mode presents a similar problem to the one we discussed yesterday if we were to choose black (which would require switching to white in dark mode). That said, because the contrast gap between accent colors in light and dark mode is far lower than the contrast gap between black and white, perhaps we can live with it for now until a more general solution for handling color selection between light/dark themes is addressed, as @flash1293 stated above.

For the sake of the immediate present, I'd say let's either stick with the EUI accent color (regardless of the fact that it will change on visualization creation depending on theme mode), or switch to same gray we use for reference lines as the default.

alexwizp commented on src/plugins/event_annotation/public/plugin.ts:
```diff
@@ -0,0 +1,39 @@
+  }
+
+  public start(): EventAnnotationPluginStart {
+    return this.eventAnnotationService!;
```
nit: you initialize that service in constructor. Extra '!'

alexwizp commented on src/plugins/event_annotation/public/plugin.ts:
```diff
@@ -0,0 +1,39 @@
+  public setup(core: CoreSetup, dependencies: SetupDependencies): EventAnnotationPluginSetup {
+    dependencies.expressions.registerFunction(manualEventAnnotation);
+    dependencies.expressions.registerFunction(eventAnnotationGroup);
+    return this.eventAnnotationService;
```
nit: I see only one method in that service, not sure that it's a legal to execute this one during setup phase. I prefer to return {} or cleanup API

mbondyra replied:
But at some point I'll add extra functionality here. I will not correct it in this PR but revisit later to make sure if it doesn't break any rules.

alexwizp commented on x-pack/plugins/lens/public/xy_visualization/annotations/config_panel/index.tsx:
```diff
@@ -0,0 +1,192 @@
+  );
+
+  return (
+    <>
```
nit: `extra <> `

mbondyra replied:
it's needed, there are two sections here.

alexwizp replied:
sorry

alexwizp commented on x-pack/plugins/lens/public/xy_visualization/annotations/helpers.tsx:
```diff
@@ -0,0 +1,227 @@
+import { defaultAnnotationColor } from '../../../../../../src/plugins/event_annotation/public';
+import { defaultAnnotationLabel } from './config_panel';
+
+const MAX_DATE = Number(new Date(8640000000000000));
```
just interesting, why we need it? 

<img width="243" alt="image" src="https://user-images.githubusercontent.com/20072247/159677033-565d4c2c-fbcf-4560-a548-9f1e66aef5b2.png">

flash1293 reviewed (commented):
It seems like there are still a few references to explicit icon placement, do we still need those?

https://github.com/elastic/kibana/blob/fc618c15f47d621a1d5143330d5f04975fe4598f/x-pack/plugins/lens/public/xy_visualization/annotations_helpers.tsx#L57

In general I really like this change, one thing we can also split out - what if on creating a new annotation it would copy over the styling from the last one in the list? That would make it easy to create lots of annotations which are similarly styled (set the styling on the first one and it will be pre-set for the rest)

mbondyra commented on x-pack/plugins/maps/public/lens/choropleth_chart/region_key_editor.tsx:
```diff
@@ -66,14 +67,14 @@ export function RegionKeyEditor(props: Props) {
   }
   return (
-    <>
+    <DimensionEditorSection>
```
This is not caused by `DimensionEditorSection` but by hacking making datasource `renderDimensionEditor` and visualization `renderDimensionEditor` look as one maaaany months ago - and it's something we want to address at some point. If you take a look at the elements from datasource, they already have the concept of sections introduced and they have the hacky solution. We spoke offline already so I hope I've explained the value of the additional element. Nevertheless, if the solution feels so hacky, I'll try other hack inside the lens and revert this one to not depend on external code.

flash1293 reviewed (approved):
LGTM, talked through a few small things offline and this is a great first version

nreese reviewed (approved):
maps changes LGTM
code review, tested in chrome

stratoula reviewed (approved):
Changes LGTM, sorry for the delay Marta. I tested everything and it works fine. Great PR, great enhancement ❤️

MichaelMarcialis commented:
> > @flash1293 about the references to icon position, it happens because the helpers are shared between reference lines and annotations so for reference lines it's still needed.
> > Copying styles from the last one in the list/ last one edited - what do you think about it @gvnmagni @MichaelMarcialis? Nevertheless, I'd address it in a separate PR.
> 
> The concept itself is really interesting, the only concern I have is that we have to find a proper way to explain what's happening to the user and I'm not sure there is a simple way to do that, I'll hear what Michael has to say

Yeah, my first instinct here is to agree with @gvnmagni. The concept is interesting, but I worry that the implementation would be non-obvious and potentially frustrating if that is not the desired behavior. I think some exploration can and should be done separately in terms of 1) users being able to establish style defaults for various aspects of Lens (including annotations) and 2) possibly consider the possibility of being able to bulk style/edit certain dimension items. Thoughts?

flash1293 commented:
Yeah, no worries, definitely something to look into later, I just kind of expected this to happen that’s why I wanted to share.
